contracts/consensus/Coconsensus.sol

```diff
@@ -0,0 +1,271 @@
+     * in blockchains.
+     */
+    mapping (bytes32 /* metachainId */ => uint256 /* blocknumber */) public blockTips;
+
+    /** Mapping of metachain id to the protocore contract address. */
+    mapping (bytes32 /* metachainId */ => ProtocoreI) public protocores;
+
+    /** Mapping of metachain id to the observers contract address. */
+    mapping (bytes32 /* metachainId */ => ObserverI) public observers;
+
+    /** Mapping of metachain id to the domain separators. */
+    mapping (bytes32 /* metachainId */ => bytes32 /* domain separator */) public domainSeparators;
+
+
+    /* Special Functions */
+
+    /**
+     * @notice Setup function does the initialization of all the mosaic
+     *         contracts on the auxiliary chain.
+     *
+     * @dev This function can be called only once.
+     */
+    function setup() public {
+
+        require(
+            selfMetachainId == bytes32(0),
+            "Coconsensus contract is already initialized."
+        );
+
+        originMetachainId = genesisOriginMetachainId;
+
+        selfMetachainId = genesisSelfMetachainId;
+
+        relativeSelfDynasty = uint256(0);
+
+        bytes32 currentMetachainId = genesisMetachainIds[SENTINEL_METACHAIN_ID];
+
+        // Loop through the genesis metachainId link list.
+        while (currentMetachainId != SENTINEL_METACHAIN_ID) {
+
+            // Setup protocore contract for the given metachain id.
+            setupProtocore(currentMetachainId);
+
+            // Setup observer contract for the given metachain id.
+            setupObserver(currentMetachainId);
+
+            // Traverse to next metachain id from the link list mapping.
+            currentMetachainId = genesisMetachainIds[currentMetachainId];
+        }
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice Observes the given block by anchoring its state root into the
+     *         corresponding observer.
+     *
+     * @param _metachainId Metachain id.
+     * @param _rlpBlockHeader RLP encoded block header.
+     *
+     * \pre `_metachainId` is not 0.
+     * \pre `_rlpBlockHeader` is not 0.
+     * \pre `_metachainId` is not `selfMetachainId`.
+     * \pre A block must exist in `blockchains` storage for the given
+     *      `_metachainid` and block number from the decoded `_rlpBlockHeader`.
+     * \pre The commit status of the block must be at least `Finalized`.
+     * \pre The status dynasty of the block less than `relativeSelfDynasty`.
+     *
+     * \post Anchors the state root in the observer contract.
+     */
+    function observeBlock(
+        bytes32 _metachainId,
+        bytes calldata _rlpBlockHeader
+    )
+        external
+    {
+        require(
+            _metachainId != bytes32(0),
+            "Metachain id must not be null."
+        );
+
+        require(
+            _rlpBlockHeader.length != 0,
+            "RLP block header must not be null."
+        );
+
+        require(
```

Comment:
This check is not harming, however, as there is no observer for the self protocore, depending on an implementation this is going to be silently ignored or failed without this check also (with current implementation it will fail). Thoughts?

p.s. \pre condition related to this one should stay.